### PR TITLE
enforce string formatting

### PR DESCRIPTION
**Pull request by mike-hunhoff**
includes changes resulting from string format enforcement.

see https://github.com/fireeye/capa/pull/468 for more information.

### CODE REVIEW

File: doc/format.md [hunk 1]
```diff
@@ -317,6 +317,7 @@ This could also be an obfuscated string, for example a stack string.
 
 The parameter is a string describing the string.
 This can be the verbatim value, or a regex matching the string.
+Verbatim values containing special characters must be surrounded by double quotes and each special character must be escaped.
```
Comment (williballenthin) [inline doc/format.md:320]: i think we need more detail here. what is a special character? lets enumerate those now before we forget.
```suggestion
Verbatim values containing special characters must be surrounded by double quotes and each special character must be escaped.
A special character is one of:
  - a backslash, which should be represented as `string: "\\"`
  - a newline or other non-space whitespace (e.g. tab, CR, LF, etc), which should be represented like `string: "\n"`
  - a double quote, which should be represented as `string: "\""`
  - a leading slash (not for a regex), which should be represented as `string: "/"`
```

File: doc/format.md [hunk 2]
```diff
@@ -328,11 +329,12 @@ Examples:
...
 ```
 - string: Firefox 64.0
+- string: "Hostname:\t\t\t%s\nIP adress:\t\t\t%s\nOS version:\t\t\t%s\n"
 - string: This program cannot be run in DOS mode.
   description: MS-DOS stub message
-- string: '{3E5FC7F9-9A51-4367-9063-A120244FBEC7}'
+- string: "{3E5FC7F9-9A51-4367-9063-A120244FBEC7}"
   description: CLSID_CMSTPLUA
-- string: '/SELECT.*FROM.*WHERE/'
+- string: /SELECT.*FROM.*WHERE/
```
Comment (williballenthin) [inline doc/format.md:337]: is this a regex? if its verbatim, needs double quotes.

Comment (williballenthin) [reply doc/format.md:337]: oh, yeah, looks like a regex, sorry.

File: doc/format.md (hunk 1 above)
Comment (williballenthin) [reply doc/format.md:320]: what about `:` and `#`?

**Review by williballenthin — CHANGES_REQUESTED** [review]
its close, but would like a bit more documentation to specify what a "special character" is and examples of how to escape them.

File: nursery/debug-build.yml [hunk 3]
```diff
@@ -7,4 +7,4 @@ rule:
   features:
     - or:
       - string: Assertion failed!
```
Comment (mr-tz) [inline nursery/debug-build.yml:9]: shouldn't strings with white space be quoted as well?

File: doc/format.md (hunk 1 above)
Comment (mr-tz) [reply doc/format.md:320]: what about spaces, see comment below?

Comment (mike-hunhoff) [reply doc/format.md]: I think the current list is comprehensive now that we enforce all strings must start/end with double quotes.

File: nursery/debug-build.yml (hunk 3 above)
Comment (mike-hunhoff) [reply nursery/debug-build.yml]: covered not that we enforce all strings start/end with double quotes.

File: doc/format.md [hunk 4]
```diff
@@ -316,7 +316,15 @@ This is probably a pointer to an ASCII or Unicode string.
 This could also be an obfuscated string, for example a stack string.
 
 The parameter is a string describing the string.
-This can be the verbatim value, or a regex matching the string.
+This can be the verbatim value or a regex matching the string.
+
+Verbatim values should be surrounded by double quotes and special characters should be escaped.
```
Comment (mr-tz) [inline doc/format.md:321]:
```suggestion
Verbatim values must be surrounded by double quotes and special characters must be escaped.
```

File: doc/format.md [hunk 5]
```diff
@@ -435,13 +444,21 @@ An ASCII or UTF-16 LE string present in the file.
 
 The parameter is a string describing the string.
 This can be the verbatim value, or a regex matching the string.
+
```
Comment (mr-tz) [inline doc/format.md:447]: To not repeat ourselves I'd prefer to refer to the above section.

**Review by mr-tz — APPROVED** [review]
`- string: "awesome!"`

**Review by Ana06 — APPROVED** [review]
LGTM

**Pull request merged**